### DISCUSSION TITLE
Feature Request: PySpark session, create dataframe from Polars dataframe? Or from DuckDBPyRelation?

### DISCUSSION THREAD

**Discussion by MarcoGorelli** [duckdb/duckdb#14868] [Ideas]
Been taking a look at the DuckDB PySpark API, and it looks really nice!

Just got a couple of requests:
- in addition to `spark.createDataFrame(pandas_df)`, would we be able to do `spark.createDataFrame(polars_df)`
- or, even better, could we create a `duckdb.experimental.spark.sql.dataframe.DataFrame` from a `DuckDBPyRelation`? Then we could use file inputs
- once we have a `duckdb.experimental.spark.sql.dataframe.DataFrame`, is there a way to get a `DuckDBPyRelation` out of it?

Use case: would love to step up support for DuckDB in [Narwhals](https://github.com/narwhals-dev/narwhals)

**Comment by Tishj** [comment]
1. Yea that's probably doable, but the goal of the pyspark API is to *match* the capabilities of the real spark API, not exceed it.
2. That's the implementation:
```py
class DataFrame:
    def __init__(self, relation: duckdb.DuckDBPyRelation, session: "SparkSession"):
        self.relation = relation
        self.session = session
        self._schema = None
        if self.relation is not None:
            self._schema = duckdb_to_spark_schema(self.relation.columns, self.relation.types)
```
3. The `relation` attribute can also be used to retrieve the wrapped Relation

**Reply by binste** [reply]
Regarding 1., I definitely see the value of prioritising features which exist in the PySpark API to make it easy to switch over. However, once that part is more complete, I'd also see the value in adding features which do not exist in the PySpark API. They could be clearly marked in the documentation and docstrings. It wouldn't hurt people coming from PySpark and could make the API moer valuable for pure duckdb users. What do you think?

**Reply by binste** [reply]
@MarcoGorelli To build on Tishj's answer, here's some working code to go from a `DuckDBPyRelation` to a `DataFrame`:

```python
import duckdb
from duckdb.experimental.spark.sql import SparkSession as session
from duckdb.experimental.spark.sql.dataframe import DataFrame
spark = session.builder.getOrCreate()
rel = spark.conn.read_parquet('https://duckdb.org/data/holdings.parquet')
df = DataFrame(rel, spark)
```

**Comment by djouallah** [comment]
for spark 4, creating a dataframe from arrow table is supported

https://github.com/apache/spark/pull/46529


https://github.com/duckdb/duckdb/pull/11349
